### ISSUE TITLE
BroadcastInDim VJP returns wrong gradient shape for non-monotonic dims

### ISSUE THREAD

**Issue by terasakisatoshi** [tensor4all/tenferro-rs#1092]
## Summary

`BroadcastInDim` の逆モード AD (`transpose_broadcast_in_dim`) は、`dims` が非単調な順序（例: 転置を模した `dims = [1, 0]`）の場合、入力軸順に戻す転置を行わずに reshape だけで済ませているため、**誤った勾配形状**が返ります。

## Affected code

`crates/tenferro-internal-ops/src/ad/structural.rs:380-426`

```rust
pub fn transpose_broadcast_in_dim(
    builder: &mut dyn PrimitiveRuleBuilder,
    cotangent_out: &[Option<LocalValueId>],
    shape: &[DimExpr],
    dims: &[usize],
    inputs: &[ValueRef<StdTensorOp>],
    ctx: &mut ShapeGuardContext,
) -> Vec<Option<LocalValueId>> {
    let (reduce_axes, needs_input_shape_restore) =
        broadcast_transpose_reduce_axes(shape, dims, inputs, ctx);
    // ...
    if needs_input_shape_restore {
        let input_rank = ctx.shape_of(&inputs[0]).len();
        let reshaped = builder.add_operation(
            StdTensorOp::Reshape {
                to_shape: DimExpr::input_shape(1, input_rank),
            },
            vec![ValueRef::Local(reduced), inputs[0].clone()],
            OperationRole::Linearized {
                active_mask: vec![true, false],
            },
        );
        Some(reshaped[0])
    } else {
        Some(reduced)
    }
}
```

`crates/tenferro-internal-ops/src/ad/structural.rs:428-466`

```rust
fn broadcast_transpose_reduce_axes(
    shape: &[DimExpr],
    dims: &[usize],
    inputs: &[ValueRef<StdTensorOp>],
    ctx: &mut ShapeGuardContext,
) -> (Vec<usize>, bool) {
    // ...
    for (input_axis, &output_axis) in dims.iter().enumerate() {
        if input_shape[input_axis] == one && output_shape[output_axis] != one {
            reduce_axes.push(output_axis);
            needs_input_shape_restore = true;
        }
    }
    // ...
}
```

## Why it is dangerous

`BroadcastInDim` は `dims` に任意の順序を取れます。例えば入力 shape `[2, 3]` に対して `dims = [1, 0]` とすると、出力 shape は `[3, 2]` になり（純粋な転置）、微分の逆伝播はその後に軸を元に戻す転置が必要です。

現在の実装は:
1. ブロードキャストされた軸を `ReduceSum` で縮約
2. 結果を `input_rank` 長の形状に reshape

しかし、`dims` が順列の場合、reshape 後の軸順は元の入力順と一致しません。`Reshape` は列優先の平坦データを単純に再分割するだけなので、転置成分が復元されず、**勾配テンソルが元の入力と異なる軸配置**になります。

これはユーザーの `grad` 呼び出しで**誤った勾配値を黙って返す**バグです。

## Expected behavior

`dims` が非単調な順列の場合、reduce 後に `dims` の逆順に相当する `Transpose` を挿入して、勾配の軸順を入力と一致させる。

## Acceptance criteria

- [ ] `transpose_broadcast_in_dim` が `dims` の順序を考慮した勾配を返す。
- [ ] 非単調 `dims` に対する有限差分または Torch oracle による回帰テストを追加。
- [ ] ブロードキャスト + 転置を組み合わせたケースのテスト。

## Severity

HIGH — silently wrong gradients

## Related

- #1084 — 広く AD ルールの panic を扱う issue（本件は panic ではなく誤った結果）

**Comment by terasakisatoshi** [comment]
## Minimal reproduction

```rust
use tenferro_ad::AdContext;
use tenferro_runtime::TracedTensor;

let ad = AdContext::builder().build().unwrap();

// x: [2, 3]
let x = TracedTensor::from_vec_col_major(
    vec![2, 3],
    (0..6).map(|i| i as f64).collect(),
);

// broadcast_in_dim with dims=[1,0] => output [3, 2] (transpose via broadcast)
let y = x.broadcast_in_dim(&[3, 2], &[1, 0]);
let loss = y.reduce_sum(&[0, 1]);

let g = ad.grad(&loss, &x).unwrap();
println!("x.shape = {:?}, grad.shape = {:?}", x.shape(), g.shape());
// Expected: grad.shape == [2, 3] with the same axis order as x.
// Bug: grad.shape is [2, 3] but axis order is transposed (values differ from finite differences).
```

### Verification against finite differences

```rust
let eps = 1e-5;
// Compare g with (loss(x+eps*e_i) - loss(x-eps*e_i)) / (2*eps)
// They will disagree for non-monotonic dims.
```
